Sources/DistributedActors/Cluster/SWIM/SWIMPeer+Actor.swift

```diff
@@ -18,16 +18,116 @@ import struct Dispatch.DispatchTime
+        sequenceNumber: SWIM.SequenceNumber
+    ) async throws -> SWIM.PingResponse {
+        guard let pingOrigin = pingOrigin as? SWIMActorShell else {
+            throw SWIMActorError.illegalPeerType("Expected origin to be \(SWIMActorShell.self) but was: \(pingOrigin)")
+        }
+
+        return try await RemoteCall.with(timeout: .nanoseconds(timeout.nanoseconds)) {
+            // FIXME: remove type cast
+            let response = try await (self as! SWIMActorShell).ping(origin: pingOrigin, payload: payload, sequenceNumber: sequenceNumber)
+            if case .nack = response {
+                throw SWIMActorError.illegalMessageType("Unexpected .nack reply to .ping message! Was: \(response)")
+            }
+            return response
+        }
+    }
+    
+    nonisolated func pingRequest(
+        target: SWIMPeer,
+        payload: SWIM.GossipPayload,
+        from pingRequestOrigin: SWIMPingRequestOriginPeer,
+        timeout: DispatchTimeInterval,
+        sequenceNumber: SWIM.SequenceNumber
+    ) async throws -> SWIM.PingResponse {
+        guard let pingRequestOrigin = pingRequestOrigin as? SWIMActorShell else {
+            throw SWIMActorError.illegalPeerType("Expected origin to be \(SWIMActorShell.self) but was: \(pingRequestOrigin)")
+        }
+
+        guard let target = target as? SWIMActorShell else {
+            throw SWIMActorError.illegalPeerType("Expected target to get \(SWIMActorShell.self) but was: \(target)")
+        }
+        
+        return try await RemoteCall.with(timeout: .nanoseconds(timeout.nanoseconds)) {
+            // FIXME: remove type cast
+            try await (self as! SWIMActorShell).pingRequest(
+                target: target,
+                pingRequestOrigin: pingRequestOrigin,
+                payload: payload,
+                sequenceNumber: sequenceNumber
+            )
+        }
+    }
+}
+
+extension SWIMActorShell: SWIMPeer {
+    nonisolated func ping(
+        payload: SWIM.GossipPayload,
+        from pingOrigin: SWIMPingOriginPeer,
+        timeout: DispatchTimeInterval,
+        sequenceNumber: SWIM.SequenceNumber,
+        onResponse: @escaping (Result<SWIM.PingResponse, Error>) -> Void
+    ) {
+        fatalError("Use async version of this API")
+    }
+
+    nonisolated func pingRequest(
+        target: SWIMPeer,
+        payload: SWIM.GossipPayload,
+        from pingRequestOrigin: SWIMPingRequestOriginPeer,
+        timeout: DispatchTimeInterval,
+        sequenceNumber: SWIM.SequenceNumber,
+        onResponse: @escaping (Result<SWIM.PingResponse, Error>) -> Void
+    ) {
+        fatalError("Use async version of this API")
+    }
+}
+
+extension SWIMActorShell: SWIMPingOriginPeer {
+    nonisolated func ack(
+        acknowledging sequenceNumber: SWIM.SequenceNumber,
+        target: SWIMPeer,
+        incarnation: SWIM.Incarnation,
+        payload: SWIM.GossipPayload
+    ) {
+        fatalError("This implementation doesn't send ack directly")
```

Comment:
Some SWIM implementations need explicit `ack`/`nack` response to trigger callbacks. This implementation doesn't need it because response is returned as the result of `ping`/`pingRequest`.

Reply:
That's good with me 👍